pkg/controller/backup_controller.go

```diff
@@ -666,17 +805,100 @@ func (c *backupController) runBackup(backup *pkgbackup.Request) error {
 		return err
 	}
 
-	if errs := persistBackup(backup, backupFile, logFile, backupStore, c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)), volumeSnapshots, volumeSnapshotContents); len(errs) > 0 {
+	if errs := persistBackupBeforeUploadComplete(backup, backupFile, logFile, backupStore, volumeSnapshots, volumeSnapshotContents); len(errs) > 0 {
 		fatalErrs = append(fatalErrs, errs...)
 	}
 
-	c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup completed")
+	snapshotInProgress := false
+	if features.IsEnabled(velerov1api.UploadProgressFeatureFlag) {
+		snapshotInProgress = checkAndUpdateItemSnapshotsProgress(backup.ItemSnapshots, logger)
+	}
+
+	if snapshotInProgress {
+		// Prepare for upload progress monitoring
+		switch {
+		case len(fatalErrs) > 0:
+			// If we had fatal errors in the backup, we will just fail here and not monitor upload progress
+			// ItemSnapshotsAttempted/Completed may be incorrect
+			backup.Status.Phase = velerov1api.BackupPhaseFailed
+		case logCounter.GetCount(logrus.ErrorLevel) > 0:
+			backup.Status.Phase = velerov1api.BackupPhaseUploadingPartialFailure
+			c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup partially failed but snapshot uploads outstanding, will complete in background")
+		default:
+			backup.Status.Phase = velerov1api.BackupPhaseUploading
+			c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Snapshot uploads outstanding, will complete in background")
+		}
+	} else {
+		updateBackupStatusForSnapshotState(backup, c.logger)
+
+		switch {
+		case len(fatalErrs) > 0:
+			backup.Status.Phase = velerov1api.BackupPhaseFailed
+		case logCounter.GetCount(logrus.ErrorLevel) > 0:
+			backup.Status.Phase = velerov1api.BackupPhasePartiallyFailed
+		default:
+			backup.Status.Phase = velerov1api.BackupPhaseCompleted
+		}
+	}
 
+	if backup.Status.Phase != velerov1api.BackupPhaseUploading &&
+		backup.Status.Phase != velerov1api.BackupPhaseUploadingPartialFailure {
+		// We're done so set the CompletionTimestamp now
+		backup.Status.CompletionTimestamp = &metav1.Time{Time: c.clock.Now()}
+		recordBackupDurationMetric(backup.Backup, c.metrics)
+		if errs := persistBackupFinal(backup, backupStore); len(errs) > 0 {
```

Comment:
Looks like whether or not upload progress enabled, there will be two uploads in object store. 
Is it possible to reduce the upload time to one for upload progress not enable case?

Reply:
The second persistBackupFinal is only called if the backup was in UploadingPhase.